### ISSUE TITLE
Feature: Utilize optimum's `BetterTransformer` for inference

### ISSUE THREAD

**Issue by michaelfeil** [UKPLab/sentence-transformers#2353]
My suggestion would be to use Optimums BetterTransformer to convert models. At least for the inference part, this can be done without any sacrifices. For training, I am not so sure, thanks to the padding masks, that raises a warning.

```python
def encode():
    model.to_bettertransformer()

def train ():
   model.reverse_bettertransformer()
```

It can be done, but this call feels sketchy :)

```python
model = SentenceTransformer(..)
model._first_module().auto_model.to_bettertransformer()
```

**Comment by tomaarsen** [comment]
The (awesome) optimum maintainers pointed me to https://github.com/huggingface/transformers/issues/26557 in `transformers`, which should automatically add a lot of the performance gain from BetterTransformers to pure `transformers` out of the box. In time, it will become more up to date than BetterTransformers from Optimum. As a result, I will be waiting for the transformers maintainers to finish this work. Until then, you can (as your post shows) manually call `to_bettertransformer()` on all `PreTrainedModel` instances within your `SentenceTransformer` instance.
I would recommend doing so if you encounter improved latency as a result.

- Tom Aarsen

**Comment by michaelfeil** [comment]
As you pointed out, most of optimum's enchancements is from `F.Scaled Dot Product Attention` and layer fusing. 
Sounds like at least `F.sdpa` is getting into transformers, which should be the major life here. https://huggingface.co/docs/transformers/perf_infer_gpu_one#bettertransformer

**Comment by michaelfeil** [comment]
Hey @tomaarsen,
I yesterday bumped infinity to `torch=2.2` & `ST=2.3.0`. Also using `torch.compile` which is now compatible with `torch._nested_tensors` of `optimum` - overall delivers around 1.7x throughput at max batch_size. This identical throughput to TEI performance of e.g. on L4.

I would expect some minor breaking changes from 2.2.2. especially when having hacked around  have been some issues using the `device`: I manually did a overwrite of `self.device = torch.device(device_name) if device_name else (torch.device("cuda") if is_gpu_avail else torch.device("cpu"))`
